lib/ddtrace/contrib/configuration/integration.rb

```diff
@@ -0,0 +1,79 @@
+module Datadog
+  module Contrib
+    module Configuration
+      # Represents an instance of a configurable integration
+      class Integration
+        attr_reader \
+          :definition,
+          :callbacks
+
+        def initialize(definition)
+          @definition = definition
+          reset
+        end
+
+        def enable!
+          @enabled = true
+        end
+
+        def disable!
+          @enabled = false
+        end
+
+        def enabled?
+          @enabled == true
+        end
+
+        def configure(*args, &block)
+          if args.first == false
+            disable!
+          elsif args.first == true
+            enable!
+          elsif args.any? || block_given?
+            add_configuration_callback(*args, &block)
+          end
+        end
+
+        # Add a configuration callback
+        def add_configuration_callback(*args, &block)
+          # Defer execution by wrapping in a proc
```

Comment:
Could you expand on why it's important to defer execution here?

Reply:
I ask not just for this PR, but also for future ourselves, browsing this method months from now :)

Reply:
Yeah, great question!

The honest answer is this need for deferment is born out of the fact that information we'd like to configure Rails instrumentation with (e.g. default service name) are not available at configuration time, but only after initialization. Therefore, if we were to apply said configuration it immediately instead of deferring, we would not end up with the values we wanted or expected.

Although this case is particular to Rails, I was thinking of the more general case where other integrations might produce similar situations: where integrations must be activated at a specific time with specific configuration. Another such example I was considering was Sinatra, which also uses Rack and is sensitive to how it's activated.

It does beg the question that if you are configuring an integration by way of another, does it always need to defer execution? The answer is probably no, and given that I'd rather avoid storing such lambdas, perhaps we need to create an alternative path where it applies configuration immediately unless explicitly configured to defer.

Reply:
In response to this feedback, I decided to add an option to define whether an integration's configuration should defer, which defaults to `false`.  When `false`, configuration is applied immediately, when `true`, configuration is put into a callback to be invoked when the integration activates it.

Reply:
I like the approach!